e2e/testcases/helm_sync_test.go

```diff
@@ -57,6 +63,58 @@ func TestPublicHelm(t *testing.T) {
+// sync the namespace scoped resources, and assign the RepoSync namespace to these resources.
+// This test will work only with following pre-requisites:
+// Google service account `e2e-test-ar-reader@stolos-dev.iam.gserviceaccount.com` is created with `roles/artifactregistry.reader` for accessing images in Artifact Registry.
+// A JSON key file is generated for this service account and stored in Secret Manager
+func TestHelmNamespaceRepo(t *testing.T) {
+	repoSyncNN := nomostest.RepoSyncNN(testNs, "rs-test")
+	nt := nomostest.New(t, ntopts.SkipMonoRepo,
+		ntopts.NamespaceRepo(repoSyncNN.Namespace, repoSyncNN.Name))
+	nt.T.Log("Update RepoSync to sync from a public Helm Chart")
+	rs := nomostest.RepoSyncObjectV1Beta1FromNonRootRepo(nt, repoSyncNN)
+	rs.Spec.SourceType = string(v1beta1.HelmSource)
+	rs.Spec.Helm = &v1beta1.HelmRepoSync{HelmBase: v1beta1.HelmBase{
+		Repo:    publicHelmRepo,
+		Chart:   publicHelmChart,
+		Auth:    configsync.AuthNone,
+		Version: publicHelmChartVersion,
+	}}
+	rs.Spec.Git = nil
+	nt.RootRepos[configsync.RootSyncName].Add(nomostest.StructuredNSPath(repoSyncNN.Namespace, repoSyncNN.Name), rs)
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Update RepoSync to sync from a public Helm Chart with cluster-scoped type")
+	nt.WaitForRepoSyncSourceError(repoSyncNN.Namespace, repoSyncNN.Name, nonhierarchical.BadScopeErrCode, "must be Namespace-scoped type")
+
+	nt.T.Log("Fetch password from Secret Manager")
+	key, err := gitproviders.FetchCloudSecret("config-sync-ci-ar-key")
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	nt.T.Log("Create secret for authentication")
+	_, err = nt.Kubectl("create", "secret", "generic", "foo", fmt.Sprintf("--namespace=%s", repoSyncNN.Namespace), "--from-literal=username=_json_key", fmt.Sprintf("--from-literal=password=%s", key))
+	if err != nil {
+		nt.T.Fatalf("failed to create secret, err: %v", err)
+	}
+	nt.T.Log("Update RepoSync to sync from a private Artifact Registry")
+	rs.Spec.Helm = &v1beta1.HelmRepoSync{HelmBase: v1beta1.HelmBase{
+		Repo:        privateARHelmRegistry,
+		Chart:       privateNSHelmChart,
+		Auth:        configsync.AuthToken,
+		Version:     privateNSHelmChartVersion,
+		ReleaseName: "test",
+		SecretRef:   v1beta1.SecretReference{Name: "foo"},
+	}}
+	rs.Spec.SourceType = string(v1beta1.HelmSource)
+	rs.Spec.Git = nil
+	nt.RootRepos[configsync.RootSyncName].Add(nomostest.StructuredNSPath(repoSyncNN.Namespace, repoSyncNN.Name), rs)
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Update RepoSync to sync from a private Helm Chart without cluster scoped resources")
+	nt.WaitForRepoSyncs(nomostest.WithRepoSha1Func(helmChartVersion(privateNSHelmChart+":"+privateNSHelmChartVersion)), nomostest.WithSyncDirectoryMap(map[types.NamespacedName]string{repoSyncNN: privateNSHelmChart}))
+	if err := nt.Validate(rs.Spec.Helm.ReleaseName+"-"+privateNSHelmChart, testNs, &appsv1.Deployment{}); err != nil {
```

Comment:
Does test cleanup work for RSync using helm? This test ends with the RepoSync configured to use helm, and my understanding is that the test functions by resetting the git repository.

Reply:
To my understand, this test uses a central git repo to control the RepoSync. When test clears resources at the end of test by resetting the root git repository, the RepoSync(for example: repo-sync.yaml) will also be removed from the repository. In this way RepoSync will be deleted from test cluster at the end of test.

Reply:
From the [test scaffolding](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/config-sync.go#L1310) it seems that the namespace repo also clears resources by resetting the git repository.

I expect the `RepoSync` will get cleaned up when the RootSync's git repository is reset. But will the resources in the RepoSync's namespace get cleaned up if the RepoSync is configured to use helm?

Reply:
Actually, I'm thinking whether we need to clean up the resources managed by RepoSync in the test scaffolding. The test is using the centralized control. In the end, the RepoSync along with its namespace will get deleted from the root repository. When the namespace is deleted, all resources in the namespace should be deleted as well.